src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -293,14 +298,27 @@ where
                                     None => {
                                         // Consume remaining rows in the builder.
                                         if let Some(data_chunk) = builder.consume_all() {
-                                            yield Self::handle_snapshot_chunk(
+                                            let chunk = Self::handle_snapshot_chunk(
                                                 data_chunk,
                                                 &mut current_pos,
                                                 &mut cur_barrier_snapshot_processed_rows,
                                                 &mut total_snapshot_processed_rows,
                                                 &pk_indices,
                                                 &self.output_indices,
                                             );
+                                            if let Some(rate_limit) = limiter.as_ref()
+                                                && chunk.cardinality() > 0
+                                            {
+                                                rate_limit
+                                                    .until_n_ready(
+                                                        NonZeroU32::new(chunk.cardinality() as u32)
+                                                            .unwrap(),
+                                                    )
+                                                    .await
+                                                    .unwrap();
+                                            }
```

Comment:
I think rate limit has to be done for the snapshot side before the snapshot and upstream are merged.

Reply:
Otherwise as @tabVersion mentions, the first barrier will be rate limited and we can't alter it..

Reply:
It's also more intuitive.

Reply:
pasting message from DM to Noel

> I am not quite sure about throttling on mv backfill, it makes the backfill slower and we cannot cancel the process because it takes longer to finish the first barrier. Also, during the process, we cannot use config change to adjust the rate limit since the barrier will be delivered after finishing the backfill. Is this desired?

Reply:
- no shuffle backfill can be modified here: https://github.com/risingwavelabs/risingwave/blob/1cc84f73cec920d52a35707620a3fd73db98b4e4/src/stream/src/executor/backfill/no_shuffle_backfill.rs#L248, basically directly apply the rate limit inside `make_snapshot_stream`.
- arrangement backfill can be modified here: https://github.com/risingwavelabs/risingwave/blob/1cc84f73cec920d52a35707620a3fd73db98b4e4/src/stream/src/executor/backfill/arrangement_backfill.rs#L246. Likewise.